### DISCUSSION TITLE
How to extend Statamic properly – Tag or Modifier?

### DISCUSSION THREAD

**Discussion by eliasu** [statamic/cms#10302] [Q&A]
Hi there, 

• I have a collection maned **Formats**
• I have a collection named **Events**

• Each **Format** contains many **Events**
• Each **Event** has multiple tags (taxonomies) like **Stage** or **Genres**

I need something like: 
```antlers
{{# get all stages used in the formats events – but since a stage can be used by many events, return an unique array #}}
{{ format.stages }} 
    {{# get all title of the stages #}}
    {{ title }} 
{{ /format.stages }}

{{# and vise versa the same for genres – a unique array for all genres used in the formats events #}}
{{ format.genres }} 
    {{# get all titles of the genres #}}
    {{ title }} 
{{ /format.genres }}
```

I assume the best idea would be to extend Statamic for this functionality? Should I use a tag or should I use a modifier or is this something for a service provider? What ist the best and most performant approach for this?

Thanks you for any help.

**Comment by ryanmitchell** [comment]
Personally I would use a query scope on the collection tag to do what you require:
https://statamic.dev/tags/collection#custom-query-scopes

**Reply by eliasu** [reply]
Hi, 
thanks for you reply. 

Did I get you right, you suggest something like this, right? (without the PHP code for now)

``` antlers
{{ collection:events format:is"Format X" query_scope="getGenres" }}
    {{# get the title of the events genre. For example "jazz" #}}
    {{ title }}}
{{ /collection:events format:is"Format X" query_scope="getGenres" }}
```

But isn't that a performance issue, since it is not iterating over all events based on a format, but iterating over all elements within the collection **events**? I mean, every format contains many events, but not all … ?

**Reply by ryanmitchell** [reply]
Within your query scope you can use the query methods - https://statamic.dev/content-queries - to limit the items returned.
